drivers/serial/uart_rpi_pico.c

```diff
@@ -39,37 +50,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+	}
+
+	if (data_reg & UART_UARTDR_FE_BITS) {
+		errors |= UART_ERROR_FRAMING;
+	}
+
+	return errors;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int uart_rpi_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data, int len)
+{
+	int tx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
+		uart_get_hw(uart_inst)->dr = tx_data[tx_len++];
+	}
+
+	return tx_len;
+}
+
+static int uart_rpi_fifo_read(const struct device *dev,
+			      uint8_t *rx_data, const int len)
+{
+	int rx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+
+	while (uart_is_readable(uart_inst) && (len - rx_len) > 0) {
+		rx_data[rx_len++] = (uint8_t)uart_get_hw(uart_inst)->dr;
+	}
+
+	return rx_len;
+}
+
+static void uart_rpi_irq_tx_enable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+
+	uart_get_hw(uart_inst)->imsc |= UART_UARTIMSC_TXIM_BITS;
+	hw_write_masked(&uart_get_hw(uart_inst)->ifls, 0, UART_UARTIFLS_TXIFLSEL_BITS);
+}
+
+static void uart_rpi_irq_tx_disable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	uart_get_hw(config->uart_dev)->imsc &= ~UART_UARTIMSC_TXIM_BITS;
+}
+
+static int uart_rpi_irq_tx_ready(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	return (uart_get_hw(config->uart_dev)->mis & UART_UARTMIS_TXMIS_BITS) == UART_UARTMIS_TXMIS_BITS;
+}
+
+static void uart_rpi_irq_rx_enable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+
+	uart_get_hw(uart_inst)->imsc |= UART_UARTIMSC_RXIM_BITS;
+	hw_write_masked(&uart_get_hw(uart_inst)->ifls, 0, UART_UARTIFLS_RXIFLSEL_BITS);
+}
+
+static void uart_rpi_irq_rx_disable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	uart_get_hw(config->uart_dev)->imsc &= ~UART_UARTIMSC_RXIM_BITS;
+}
+
+static int uart_rpi_irq_tx_complete(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	return !!(uart_get_hw(config->uart_dev)->fr & UART_UARTFR_TXFE_BITS);
+}
+
+static int uart_rpi_irq_rx_ready(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	return (uart_get_hw(config->uart_dev)->mis & UART_UARTMIS_RXMIS_BITS) == UART_UARTMIS_RXMIS_BITS;
+}
+
+static void uart_rpi_irq_err_enable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	uart_get_hw(config->uart_dev)->imsc |= (UART_UARTIMSC_OEIM_BITS |
+											UART_UARTIMSC_BEIM_BITS |
+											UART_UARTIMSC_PEIM_BITS |
+											UART_UARTIMSC_FEIM_BITS |
+											UART_UARTIMSC_RTIM_BITS);
```

Comment:
looks like formatting needs to be improved here

Reply:
Yes, I formatted it.